src/parser/warlock/demonology/modules/features/SummonDemonicTyrant.js

```diff
@@ -0,0 +1,107 @@
+import React from 'react';
+
+import Analyzer , {SELECTED_PLAYER} from 'parser/core/Analyzer';
+import SPELLS from 'common/SPELLS';
+import Events from 'parser/core/Events';
+
+import StatisticBox from 'interface/others/StatisticBox';
+import SpellIcon from 'common/SpellIcon';
+
+import DemoPets from '../pets/DemoPets';
+import { isWildImp, isDreadstalker, isVilefiend, isPermanentPet } from '../pets/helpers';
+
+class SummonDemonicTyrant extends Analyzer {
+  static dependencies = {
+    demoPets: DemoPets,
+  };
+
+  _petsPerCast = [];
+
+  _dreadstalkersPerCast = [];
+  _impsPerCast = [];
+  _vilefiendsPerCast = [];
+  _permPetsPerCast = [];
+
+  constructor(...args){
+    super(...args);
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.SUMMON_DEMONIC_TYRANT), this.summonDemonicTyrantCast);
+  }
+
+  summonDemonicTyrantCast(event) {
```

Comment:
While I was working on the DemoPets module, they changed yet again what Demonic Tyrant empowers. As far as I know, these days, Demonic Tyrant empowers *all* pets in one way or another (and talking about "controllable pets" is weird as well since you don't really control Imps or Dreadstalkers any more than the random pets summoned with Inner Demons for example, unless by "controllable" you mean that you control when you summon them). With that said, I think some refactor here and in the display would be helpful. 

First of all, simplifying the counting of the pets could be done by utilizing the `summonedBy` property of the pet (https://github.com/WoWAnalyzer/WoWAnalyzer/blob/679bef449d55299edef62afe65f85925337be81d/src/parser/warlock/demonology/modules/pets/TimelinePet.js#L33). This property contains the ID of the ability that summoned the pet from player's perspective, Hand of Gul'dan, Call Dreadstalkers, but also Inner Demons or Nether Portal.

Something along the lines of 
```javascript
const pets = this.demoPets.currentPets;

const countsPerCast = {};
pets.forEach(pet => {
  countsPerCast[pet.summonedBy] = (countsPerCast[pet.summonedBy] || 0) + 1;
});
this._petsPerCast.push(countsPerCast);
```

Since this would make the rendered table wider, I either suggest transposing it (make the columns into rows and rows into columns), that would put different demons in rows and casts in columns (that might not be as bad since you usually manage to get just a few casts per fight, if you don't count things like G'huun).

Second option is to drop showing *individual* casts and average the pets too. So it would show like "average pets empowered: 7.12", "average Wild Imps: 4.86", "average Dreadstalkers: 2.00" etc.

Reply:
As per your suggestion, I've modified it to track the number of demons summoned from each source per cast of Summon Demonic Tyrant, then present the average for each summon source.

![image](https://user-images.githubusercontent.com/6363390/51002514-814d7b00-1501-11e9-843d-d31b352c821c.png)

Reply:
Note the repetition of the Felguard spell icon in the screenshot is for Summon Felguard and for Grimoire: Felguard. These are SpellIcons so it's clarified on mouseover.